From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kenton Varda <kenton@cloudflare.com>
Date: Wed, 4 Dec 2024 22:36:05 -0600
Subject: Add ValueSerializer::SetTreatProxiesAsHostObjects().

Previously, ValueSerializer would always refuse to serialize Proxy objects. This commit gives the embedder the option to handle them as host objects.

Similar to the previous patch adding `SetTreatFunctionsAsHostObjects()`, this is intended for use in an RPC system, where an arbitrary object can be "serialized" by replacing it with a stub which, when invoked, performs an RPC back to the originating isolate in order to access the original object there.

Signed-off-by: James M Snell <jsnell@cloudflare.com>

diff --git a/include/v8-value-serializer.h b/include/v8-value-serializer.h
index 141f138e08de849e3e02b3b2b346e643b9e40c70..bdcb2831c55e21c6d511f56dfc79a5076871f05a 100644
--- a/include/v8-value-serializer.h
+++ b/include/v8-value-serializer.h
@@ -204,6 +204,15 @@ class V8_EXPORT ValueSerializer {
    */
   void SetTreatFunctionsAsHostObjects(bool mode);
 
+  /**
+   * Indicate whether to treat Proxies as host objects,
+   * i.e. pass them to Delegate::WriteHostObject. This should not be
+   * called when no Delegate was passed.
+   *
+   * The default is not to treat Proxies as host objects.
+   */
+  void SetTreatProxiesAsHostObjects(bool mode);
+
   /**
    * Write raw data in various common formats to the buffer.
    * Note that integer types are written in base-128 varint format, not with a
diff --git a/src/api/api.cc b/src/api/api.cc
index 6682ce73a813f6e795c442534f6303964e1f9a61..2aaae63b805d221bc92e7b1e56a383ce02a6cd37 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -3325,6 +3325,10 @@ void ValueSerializer::SetTreatFunctionsAsHostObjects(bool mode) {
   private_->serializer.SetTreatFunctionsAsHostObjects(mode);
 }
 
+void ValueSerializer::SetTreatProxiesAsHostObjects(bool mode) {
+  private_->serializer.SetTreatProxiesAsHostObjects(mode);
+}
+
 Maybe<bool> ValueSerializer::WriteValue(Local<Context> context,
                                         Local<Value> value) {
   auto i_isolate = reinterpret_cast<i::Isolate*>(context->GetIsolate());
diff --git a/src/objects/value-serializer.cc b/src/objects/value-serializer.cc
index 5fe04fecb1cc6bc98b8d3571a540c418c3fa85f7..ca0166b2c7fe791034c900fc274c595f7908cc4e 100644
--- a/src/objects/value-serializer.cc
+++ b/src/objects/value-serializer.cc
@@ -338,6 +338,10 @@ void ValueSerializer::SetTreatFunctionsAsHostObjects(bool mode) {
   treat_functions_as_host_objects_ = mode;
 }
 
+void ValueSerializer::SetTreatProxiesAsHostObjects(bool mode) {
+  treat_proxies_as_host_objects_ = mode;
+}
+
 void ValueSerializer::WriteTag(SerializationTag tag) {
   uint8_t raw_tag = static_cast<uint8_t>(tag);
   WriteRawBytes(&raw_tag, sizeof(raw_tag));
@@ -609,7 +613,12 @@ Maybe<bool> ValueSerializer::WriteJSReceiver(
   InstanceType instance_type = receiver->map()->instance_type();
   if (IsCallable(*receiver)) {
     if (treat_functions_as_host_objects_) {
-      return WriteHostObject(Cast<JSObject>(receiver));
+      return WriteHostObject(receiver);
+    }
+    return ThrowDataCloneError(MessageTemplate::kDataCloneError, receiver);
+  } else if (instance_type == JS_PROXY_TYPE) {
+    if (treat_proxies_as_host_objects_) {
+      return WriteHostObject(receiver);
     }
     return ThrowDataCloneError(MessageTemplate::kDataCloneError, receiver);
   } else if (IsSpecialReceiverInstanceType(instance_type) &&
@@ -1268,7 +1277,7 @@ Maybe<bool> ValueSerializer::WriteSharedObject(
   return ThrowIfOutOfMemory();
 }
 
-Maybe<bool> ValueSerializer::WriteHostObject(DirectHandle<JSObject> object) {
+Maybe<bool> ValueSerializer::WriteHostObject(DirectHandle<JSReceiver> object) {
   WriteTag(SerializationTag::kHostObject);
   if (!delegate_) {
     isolate_->Throw(*isolate_->factory()->NewError(
diff --git a/src/objects/value-serializer.h b/src/objects/value-serializer.h
index 2d0a746c90522500fcf17d73f3024512b1bb5481..7ec442c5e002e8a7e00c7d0ea5dc3e77744bb55d 100644
--- a/src/objects/value-serializer.h
+++ b/src/objects/value-serializer.h
@@ -111,6 +111,15 @@ class ValueSerializer {
    */
   void SetTreatFunctionsAsHostObjects(bool mode);
 
+  /*
+   * Indicate whether to treat Proxies as host objects,
+   * i.e. pass them to Delegate::WriteHostObject. This should not be
+   * called when no Delegate was passed.
+   *
+   * The default is not to treat Proxies as host objects.
+   */
+  void SetTreatProxiesAsHostObjects(bool mode);
+
  private:
   // Managing allocations of the internal buffer.
   Maybe<bool> ExpandBuffer(size_t required_capacity);
@@ -161,8 +170,7 @@ class ValueSerializer {
 #endif  // V8_ENABLE_WEBASSEMBLY
   Maybe<bool> WriteSharedObject(DirectHandle<HeapObject> object)
       V8_WARN_UNUSED_RESULT;
-  Maybe<bool> WriteHostObject(DirectHandle<JSObject> object)
-      V8_WARN_UNUSED_RESULT;
+  Maybe<bool> WriteHostObject(DirectHandle<JSReceiver> object) V8_WARN_UNUSED_RESULT;
 
   /*
    * Reads the specified keys from the object and writes key-value pairs to the
@@ -195,6 +203,7 @@ class ValueSerializer {
   bool has_custom_host_objects_ = false;
   bool treat_array_buffer_views_as_host_objects_ = false;
   bool treat_functions_as_host_objects_ = false;
+  bool treat_proxies_as_host_objects_ = false;
   bool out_of_memory_ = false;
   Zone zone_;
   uint32_t version_;
